### DISCUSSION TITLE
What are the differences and relationships between TimeoutOptions and defaultTimeout of AbstractRedisClient?

### DISCUSSION THREAD

**Discussion by yangkangquan123** [redis/lettuce#2913] [Q&A]
I found that AbstractRedisClient has a defaultTimeout, and I saw a TimeoutOptions that also set the timeout. What is the difference and connection between the two timeout settings?

**Comment by tishun** [comment]
See similar discussion #2614 

The `AbstractRedisClient.setDefaultTimeout(Duration)` is deprecated as per the documentation:
```java
    /**
     * Set the default timeout for connections created by this client. The timeout applies to connection attempts and
     * non-blocking commands.
     *
     * @param timeout default connection timeout, must not be {@code null}.
     * @since 5.0
     * @deprecated since 6.2, use {@link RedisURI#getTimeout()} to control timeouts.
     */
    @Deprecated
    public void setDefaultTimeout(Duration timeout) {
```

Both this method and the `RedisURI.setTimeout(Duration)` (which is the recommended way to set a default timeout) set the maximum amount of time for the client to establish connections or receive replies for synchronous commands. If none is set then the driver applies a 60 second timeout by default.

The [ClientOptions](https://lettuce.io/core/release/reference/#client-options) allow to override that with more fine-grained settings using the `TimeoutOptions` including how to determine a timeout using a `TimeoutSource`. There is an [example of how TimeoutSource can be used](https://lettuce.io/core/release/reference/#faq.timeout.blpop) in the Lettuce reference guide.
